pypeit/scripts/coadd_datacube.py

```diff
@@ -0,0 +1,412 @@
+            # Obtain a rough guess of the highest S/N cube
+            sn_img = whitelight_Imgs[:, :, ff]*msk / (whitelight_Errs[:, :, ff] + (whitelight_Errs[:, :, ff] == 0))
+            maxval = np.max(sn_img)
+            if maxval > bestmax:
+                bestmax = maxval
+                ref_idx = ff
+
+        msgs.info("Calculating the relative spatial translation of each cube (reference cube = {0:d})".format(ref_idx+1))
+        # Calculate the image offsets - check the reference is a zero shift
+        ra_shift_ref, dec_shift_ref = calculate_image_offset(whitelight_Imgs[:, :, ref_idx], whitelight_Imgs[:, :, ref_idx])
+        for ff in range(numfiles):
+            # Don't correlate the reference image with itself
+            if ff == ref_idx:
+                continue
+            # Calculate the shift
+            ra_shift, dec_shift = calculate_image_offset(whitelight_Imgs[:, :, ff], whitelight_Imgs[:, :, ref_idx])
+            # Convert to reference
+            ra_shift -= ra_shift_ref
+            dec_shift -= dec_shift_ref
+            # Convert pixel shift to degress shift
+            ra_shift *= dspat/cosdec
+            dec_shift *= dspat
+            msgs.info("Image shift of cube {0:d}: RA, DEC (arcsec) = {1:+0.3f}, {2:+0.3f}".format(ff+1, ra_shift*3600.0, dec_shift*3600.0))
+            # Apply the shift
+            all_ra[all_idx == ff] += ra_shift
+            all_dec[all_idx == ff] += dec_shift
+
+    # Generate a master WCS to register all frames
+    coord_min = [np.min(all_ra), np.min(all_dec), np.min(all_wave)]
+    coord_dlt = [dspat, dspat, dwv]
+    masterwcs = generate_masterWCS(coord_min, coord_dlt)
+
+    # Generate the output binning
+    if combine:
+        numra = int((np.max(all_ra)-np.min(all_ra)) * cosdec / dspat)
+        numdec = int((np.max(all_dec)-np.min(all_dec))/dspat)
+        numwav = int((np.max(all_wave)-np.min(all_wave))/dwv)
+        xbins = np.arange(1+numra)-0.5
+        ybins = np.arange(1+numdec)-0.5
+        zbins = np.arange(1+numwav)-0.5
+    else:
+        # TODO :: This is KCWI specific - probably should put this in the spectrograph file, or just delete it.
+        slitlength = int(np.round(np.median(slits.get_slitlengths(initial=True, median=True))))
+        xbins = np.arange(1 + 24) - 12.0 - 0.5
+        ybins = np.linspace(np.min(minmax[:, 0]), np.max(minmax[:, 1]), 1+slitlength) - 0.5
+        zbins = np.arange(1+int(round((np.max(waveimg)-wave0)/dwv))) - 0.5
+
+    # Make the cube
+    msgs.info("Generating datacube")
+    # TODO :: Need to include variance weights and make a variance cube
+    if combine:
+        pix_coord = masterwcs.wcs_world2pix(all_ra, all_dec, all_wave * 1.0E-10, 0)
+        hdr = masterwcs.to_header()
+    else:
+        pix_coord = wcs.wcs_world2pix(np.vstack((all_ra, all_dec, all_wave*1.0E-10)).T, 0)
+        hdr = wcs.to_header()
+    datacube, edges = np.histogramdd(pix_coord, bins=(xbins, ybins, zbins), weights=all_sci*all_ivar)
```

Comment:
You are computing an inverse variance weighted datacube here. While that is formally optimal, In my experience co-adding/extracting images with noisy individual pixel weights is rarely a good idea.  For co-adding images, I've found it is usually preferable to work with a single weight per exposure, or a smoothly varying weight function (in the spectral direction), the describes the S/N ratio as a function of wavelength.  It is always better to use a model for the inverse variance, which you don't have here, since you don't have a model of your objects and have only a model of the sky. So I think ivar must be that computed from the image which is noisy. This will also be biased in the low-count regime and I think produce coadded cubes with excess noise.  I think it might be useful to compute a "vanilla" coadded cube without these noisy weights, i.e. using individual exposure weights, or using weights that are smoother than the ivar image.

Reply:
Hi @rcooke-ast the reason I adopt a smoothly varying S/N ratio as a function of wavelength is to deal exactly with this non-overlapping issue, i.e. if one were to somehow combine two setups with different wavelength coverage. The problem is you need to extract an object to get a smoothly varying S/N ratio. 

The problem with what you are doing in my opinion is that suppose that the transparency changes but the sky stays the same, which can often be the case. Then you are going to weight to exposures the same, whereas you should be waiting the one with higher transparency more. 

Implementing correct S/N ratio weights, and also smoothly varying wavelength weights requires an estimate of the S/N for the same reference object on each image. This sounds challenging before you have the full cube. One approach could be to do a preliminary stack, perhaps uniform weigted, or with approximate single exposure weights, and then create the cube, use object finding to locate a bright object in the preliminary cube, estimate S/N for that object in each exposure, and then recompute the cube. That is in my opinion the "optimal" way to do it.